src/aliceVision/keyframe/KeyframeSelector.cpp

```diff
@@ -25,452 +25,711 @@ namespace keyframe {
-    mediaInfo.spec.attribute("Make",  _cameraInfos[mediaIndex].brand);
-    mediaInfo.spec.attribute("Model", _cameraInfos[mediaIndex].model);
-    mediaInfo.spec.attribute("Exif:BodySerialNumber", std::to_string(getRandomInt())); // TODO: use Exif:OriginalRawFileName instead
-    mediaInfo.spec.attribute("Exif:FocalLength", _cameraInfos[mediaIndex].focalLength);
-  }
-
-  // iteration process
-  _keyframeIndexes.clear();
-  std::size_t currentFrameStep = _minFrameStep + 1; // start directly (dont skip minFrameStep first frames)
-  
-  for(std::size_t frameIndex = 0; frameIndex < _framesData.size(); ++frameIndex)
-  {
-    ALICEVISION_LOG_INFO("frame : " << frameIndex);
-    bool frameSelected = true;
-    auto& frameData = _framesData.at(frameIndex);
-    frameData.mediasData.resize(_feeds.size());
-
-    for(std::size_t mediaIndex = 0; mediaIndex < _feeds.size(); ++mediaIndex)
-    {
-      ALICEVISION_LOG_DEBUG("media : " << _mediaPaths.at(mediaIndex));
-      auto& feed = *_feeds.at(mediaIndex);
-
-      if(frameSelected) // false if a camera of a rig is not selected
-      {
-        if(!feed.readImage(image, queryIntrinsics, currentImgName, hasIntrinsics))
-        {
-          ALICEVISION_LOG_ERROR("Cannot read frame '" << currentImgName << "' !");
-          throw std::invalid_argument("Cannot read frame '" + currentImgName + "' !");
+    /**
+     * To respect the _minFrameStep, _maxFrameStep and _maxOutFrames constraints as much as possible:
+     * - if _maxOutFrames is set and the current step is too small to sample over the entire sequence,
+     *   the step should be increased;
+     * - if _maxOutFrames is set and the adjusted step became too big and does not respect _maxFrameStep anymore,
+     *   the step should be set to _maxFrameStep - in that case, _maxOutFrames might be reached before the end of
+     *   the sequence
+     */
+    if (_maxOutFrames > 0 && nbFrames / _maxOutFrames > step) {
+        step = (nbFrames / _maxOutFrames) + 1;  // + 1 to prevent ending up with more than _maxOutFrame selected frames
+        if (_maxFrameStep > 0 && step > _maxFrameStep) {
+            step = _maxFrameStep;
         }
+    }
 
-        // compute sharpness and sparse distance
-        if(!computeFrameData(image, frameIndex, mediaIndex, tileSharpSubset))
-        {
-          frameSelected = false;
-        }
-      }
+    for (unsigned int id = 0; id < nbFrames; id += step) {
+        ALICEVISION_LOG_DEBUG("Selecting frame with ID " << id);
+        _selectedKeyframes.push_back(id);
+        if (_maxOutFrames > 0 && _selectedKeyframes.size() >= _maxOutFrames)
+            break;
+    }
+}
 
-      feed.goToNextFrame();
+void KeyframeSelector::processSmart(const float pxDisplacement, const std::size_t rescaledWidth,
+                                    const std::size_t sharpnessWindowSize, const std::size_t flowCellSize)
+{
+    _selectedKeyframes.clear();
+    _selectedFrames.clear();
+
+    /**
```

Comment:
this kind of high-level explanation could also be in the function documentation, this way it would be accessible from the documentation generated by doxygen

Reply:
Done.